src/RecursiveStagesService.php

```diff
@@ -0,0 +1,166 @@
+<?php
+
+namespace SilverStripe\Versioned;
+
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\SS_List;
+
+/**
+ * Functionality for detecting the need of publishing nested objects owned by common parent / ancestor object
+ */
+class RecursiveStagesService implements RecursiveStagesInterface
+{
+    use Injectable;
+
+    /**
+     * Strong ownership uses 'owns' configuration to determine relationships
+     */
+    public const OWNERSHIP_STRONG = 'strong';
+
+    /**
+     * Weak ownership uses 'cascade_duplicates' configuration to determine relationships
+     */
+    public const OWNERSHIP_WEAK = 'weak';
+
+    /**
+     * Determine if content differs on stages including nested objects
+     *
+     * @param DataObject $object
+     * @param string $mode
+     * @return bool
+     */
+    public function stagesDifferRecursive(DataObject $object, string $mode): bool
+    {
+        if (!$object->exists()) {
+            return false;
+        }
+
+        $items = [$object];
+
+        // compare existing content
+        while ($item = array_shift($items)) {
+            if ($this->checkNeedPublishingItem($item)) {
+                return true;
+            }
+
+            $relatedObjects = $this->findOwnedObjects($item, $mode);
+            $items = array_merge($items, $relatedObjects);
+        }
+
+        // compare deleted content
+        $draftIdentifiers = $this->findOwnedIdentifiers($object, $mode, Versioned::DRAFT);
+        $liveIdentifiers = $this->findOwnedIdentifiers($object, $mode, Versioned::LIVE);
+
+        return $draftIdentifiers !== $liveIdentifiers;
+    }
+
+    /**
+     * Find all identifiers for owned objects
+     *
+     * @param DataObject $object
+     * @param string $mode
+     * @param string $stage
+     * @return array
+     */
+    protected function findOwnedIdentifiers(DataObject $object, string $mode, string $stage): array
+    {
+        $ids = Versioned::withVersionedMode(function () use ($object, $mode, $stage): array {
+            Versioned::set_stage($stage);
+
+            $object = DataObject::get_by_id($object->ClassName, $object->ID);
```

Comment:
It feels a little odd that you're overriding the object with itself. I gather that the intention is that the assignment of the second object reference is that it's the object for the given stage. With that in mind, please make this change which makes it a little easier to see what's happening at a glance

```suggestion
        $ids = Versioned::withVersionedMode(function () use ($object, $mode, $stage): array {
            Versioned::set_stage($stage);

            $stagedObject = DataObject::get_by_id($object->ClassName, $object->ID);
```